src/upsample.jl

```diff
@@ -64,235 +64,163 @@ function ChainRulesCore.rrule(::typeof(upsample_nearest), x::AbstractArray, s::T
+@inline function compute_source_index_and_lambda(
+    ratio, # 0 < ratio < 1
+    output_index,
+    input_size,
+    output_size
+)
+    real_input_index = ratio*output_index
+    input_index0 = floor(Int, real_input_index) # typecast to int was here in C++
+    offset = (input_index0 < input_size - 1) ? 1 : 0
+    input_index1 = input_index0 + offset
+    lambda1 = real_input_index - input_index0
+    lambda0 = 1 - lambda1
+    return input_index0, input_index1, lambda0, lambda1
 end
 
 """
-    adjoint_of_idx(idx::Vector{<:Integer})
-
-# Arguments
-- `idx`: a vector of indices from which you want the adjoint.
+    upsample_bilinear(x::AbstractArray{T,4}, scale::NTuple{2,Real}=(1,1); size::Union{Nothing,NTuple{2,Integer}}=nothing)
 
-# Outputs
--`idx_adjoint`: index that inverses the operation `x[idx]`.
+Upsamples the first 2 dimensions of the array `x` by the upsample factors stored in `scale`,
+using bilinear interpolation.
 
-# Explanation
-Determines the adjoint of the vector of indices `idx`, based on the following assumptions:
-* `idx[1] == 1`
-* `all(d in [0,1] for d in diff(idx))`
-The adjoint of `idx` can be seen as an inverse operation such that:
+The size of the output is equal to
+`(scale[1]*S1, scale[2]*S2, S3, S4)`, where `S1, S2, S3, S4 = size(x)`.
 
+Examples:
 ```julia
-x = [1, 2, 3, 4, 5]
-idx = [1, 2, 2, 3, 4, 4, 5]
-idx_adjoint = adjoint_of_idx(idx)
-@assert x[idx][idx_adjoint] == x
+upsample_bilinear(x, (2, pi)) # real scaling factors are allowed
+upsample_bilinear(x; size=(64,64)) # note the semicolon, size is a keyword argument
 ```
-The above holds as long as `idx` contains every index in `x`.
+Currently only 2d upsampling is supported.
 """
-function adjoint_of_idx(idx::Vector{Int})
-    d = trues(length(idx))
-    d[2:end] .= diff(idx)
-    idx_adjoint = findall(d)
-    return idx_adjoint
+function upsample_bilinear(x::AbstractArray{T,4}, scale::NTuple{2,Real}=(1,1); size::Union{Nothing,NTuple{2,Integer}}=nothing) where T
+    w,h,c,n = Base.Base.size(x)
+    if size===nothing
+        out_w = floor(Int, scale[1]*w)
+        out_h = floor(Int, scale[2]*h)
+    else
+        out_w, out_h = size
+    end
+    y = Array{T,4}(undef, out_w, out_h, c, n)
+    return upsample_bilinear_whcn!(y, x)
 end
 
-function get_newsize(sz, k)
-    return ntuple(i -> i <= length(k) ? sz[i]*k[i] : sz[i], length(sz))
+upsample_bilinear(x, scale::Real; size=nothing) = upsample_bilinear(x, (scale,scale); size=size)
+
+# this is the core function which works on arrays of arbitrary size
+# the implementation is a translation of https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cpu/UpSampleMoreKernel.cpp
+# which implements open-cv style linear interpolation / upsampling
+# for simplicity, corners are aligned and all logic for other behaviour has been stripped
+# - whcn because there is also a cwhn implementation
+# - the function is parallelized using @threads
+# - RGB types could be supported via reinterpreting
```

Comment:
Can we make them work out of the box

Reply:
Well... yes and no. Leaving the batch dim out, assuming the data comes in form `Matrix{RGB}` we could reinterpret to `Array{UInt8, 3}` with size (3,W,H). Then we would have to permute the dims to WHC, call the function with that and convert back; rather inefficient. There's also a lazy path using PermutedDimsView, I don't know how that is performance-wise. `@avx` could *maybe* be used to rearrange the loops in the compiled code automatically, but it's also super easy to provide a function which works on cwhn directly; just switch loop order and `getindex` calls, done.

I would like to leave this for later, as this is probably no deep learning use case.

Reply:
It is a pity that RGB types don't support scalar multiplication, otherwise this could be very simple:
```
julia> using ImageCore

julia> 0.5 * RGB(0.2, 0.3, 0.4) + 0.5 * RGB(0.6, 0.5, 0.4)
ERROR: MethodError: no method matching *(::Float64, ::RGB{Float64})
```
With reinterpret, as you say, they would end up as the first dimension not the 3rd. PermutedDimsArray could move this, without a copy, might not be inefficient... seems fine to leave until someone needs it, though. Nowhere else does this package build anything specially for RGB.

Reply:
Well, we intend to take a differentiable programming view. Many times we support RGB implicitly. Maybe we should be adding specific tests for this? Thats somewhat separate though

Reply:
> It is a pity that RGB types don't support scalar multiplication, otherwise this could be very simple

Aaah wait, one could in theory reinterpret to an SVector, then it should work with setting `C=1`.

> And allowing the user to specify which dimensions are channel/batch-like would add complexity.

That would maybe require implementing sth like [NamedDims.jl](https://github.com/invenia/NamedDims.jl) and dispatch on the dimension names or similar, so quite a big overhaul. I would love to see sth like that for the whole DL ecosystem so that I get get rid of all transposes during data preparation. Maybe when I retire :rofl:

Reply:
I'd like to leave this for later, therefore resolving.